flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/GroupedPartitionWriter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.filesystem;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.io.OutputFormat;
+
+/**
+ * {@link PartitionWriter} for grouped dynamic partition inserting. It will create a new format
+ * when partition changed.
+ *
+ * @param <T> The type of the consumed records.
+ */
+@Internal
+public class GroupedPartitionWriter<T> implements PartitionWriter<T> {
+
+	private final Context<T> context;
+	private final PathGenerator pathGenerator;
+	private final PartitionComputer<T> computer;
+	private final PartitionPathMaker maker;
+
+	private OutputFormat<T> currentFormat;
+	private String currentPartition;
+
+	public GroupedPartitionWriter(
+			Context<T> context,
+			PathGenerator pathGenerator,
+			PartitionComputer<T> computer,
+			PartitionPathMaker maker) {
+		this.context = context;
+		this.pathGenerator = pathGenerator;
+		this.computer = computer;
+		this.maker = maker;
+	}
+
+	@Override
+	public void write(T in) throws Exception {
+		String partition = maker.makePartitionPath(computer.makePartitionValues(in));
+		if (currentPartition == null || !partition.equals(currentPartition)) {
```

Comment:
unnecessary null check

Reply:
For grouped input, at the first, the `currentPartition` is null. And then, we need close previous format.

Reply:
partition.equals(null) will be false, no?

Reply:
Yes, you are right.